core/src/stored/reserve.cc

```diff
@@ -354,10 +305,83 @@ static bool UseDeviceCmd(JobControlRecord* jcr)
     Dmsg1(debuglevel, ">dird: %s", dir->msg);
   }
 
-  ReleaseReserveMessages(jcr);
+  ClearReserveMessages(jcr);
   return ok;
 }
 
+/**
+ * We get the following type of information:
+ *
+ * use storage=xxx media_type=yyy pool_name=xxx pool_type=yyy append=1 copy=0
+ * strip=0 use device=zzz use device=aaa use device=bbb use storage=xxx
+ * media_type=yyy pool_name=xxx pool_type=yyy append=0 copy=0 strip=0 use
+ * device=bbb
+ */
+static bool UseDeviceCmd(JobControlRecord* jcr)
+{
+  PoolMem StoreName, dev_name, media_type, pool_name, pool_type;
+  BareosSocket* dir = jcr->dir_bsock;
+  int32_t append;
+  bool ok = true;
+  int32_t Copy, Stripe;
+
+  /* If there are multiple devices, the director sends us
+   * use_device for each device that it wants to use. */
+  do {
+    Dmsg1(debuglevel, "<dird: %s", dir->msg);
+    auto res
+        = sscanf(dir->msg, use_storage, StoreName.c_str(), media_type.c_str(),
+                 pool_name.c_str(), pool_type.c_str(), &append, &Copy, &Stripe);
+
+    if (res != 7) { ok = false; }
+
+    if (!ok) { break; }
+
+    auto& storages = jcr->sd_impl->dirstores;
+
+    UnbashSpaces(StoreName);
+    UnbashSpaces(media_type);
+    UnbashSpaces(pool_name);
+    UnbashSpaces(pool_type);
+    auto& storage
+        = storages.emplace_back(append, StoreName.c_str(), media_type.c_str(),
+                                pool_name.c_str(), pool_type.c_str());
+
+    // Now get all devices
+    while (dir->recv() >= 0) {
+      Dmsg1(debuglevel, "<dird device: %s", dir->msg);
+      ok = sscanf(dir->msg, use_device, dev_name.c_str()) == 1;
+      if (!ok) { break; }
+      UnbashSpaces(dev_name);
+      storage.device_names.emplace_back(dev_name.c_str());
+    }
+
+  } while (ok && dir->recv() >= 0);
+
+  if (!ok) {
+    PmStrcpy(jcr->errmsg, dir->msg);
+    UnbashSpaces(jcr->errmsg);
+    Jmsg(jcr, M_FATAL, 0, T_("Failed command: %s\n"), jcr->errmsg);
+    dir->fsend(BAD_use, jcr->errmsg);
+    Dmsg1(debuglevel, ">dird: %s", dir->msg);
+    return false;
+  }
+
+  if (me->just_in_time_reservation && append) {
+    jcr->sd_impl->dcr = nullptr;  // signal rest of storage daemon that
+                                  // device is not used.
+    PmStrcpy(dev_name, "I am lying, its not actually reserved :)");
```

Comment:
This string is printed into the joblog without spaces.
Also, it probably would make more sense to put something different here.
`just_in_time_reservation_device`

Reply:
Also, as a bareos user I would like to see the device that was really chosen when data comes in.

Reply:
If no reservation was done at all because no data was ever sent, we also should have a line that says something like `No data was written so no device was reserved.`